src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,462 @@
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = -restricted_gibbs_sampling(true);
+
+  double log_ratio_likelihoods = 0;
+  double log_ratio_prior_prob = 0;
+  compute_log_ratio_like_and_prior(first_random_idx, second_random_idx, true,
+                                   log_ratio_prior_prob,
+                                   log_ratio_likelihoods);
+
+  const double AcRa =
+      std::min(1.0, std::exp(log_ratio_transition_prob + log_ratio_prior_prob +
+                             log_ratio_likelihoods));
+  if (accepted_proposal(AcRa)) {
+    proposal_update_allocations(first_random_idx, second_random_idx, true);
+  }
+}
+
+void SplitAndMergeAlgorithm::merge(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = 0;
+  // Fake Gibbs Sampling in order to compute the transition probability
+  for (unsigned int k = 0; k < restricted_gs_data_idx.size(); k++) {
+    restricted_gs_unique_values[allocations_restricted_gs[k]]->remove_datum(
+        restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+        update_hierarchy_params());
+
+    Eigen::VectorXd logprobas(2);
+
+    if (restricted_gs_unique_values[0]->get_card() >= 1 and
+        restricted_gs_unique_values[1]->get_card() >= 1) {
+      logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+      logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+    } else {
+      if (restricted_gs_unique_values[1]->get_card() == 0) {
+        logprobas(1) = restricted_gs_unique_values[1]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      } else {
+        logprobas(0) = restricted_gs_unique_values[0]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      }
+    }
+    if (allocations[restricted_gs_data_idx[k]] ==
+        allocations[first_random_idx]) {
+      allocations_restricted_gs[k] = 0;
+      restricted_gs_unique_values[0]->add_datum(
+          restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+          update_hierarchy_params());
+      log_ratio_transition_prob += log(stan::math::softmax(logprobas)(0));
+    } else {
+      allocations_restricted_gs[k] = 1;
+      restricted_gs_unique_values[1]->add_datum(
+          restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+          update_hierarchy_params());
+      log_ratio_transition_prob += log(stan::math::softmax(logprobas)(1));
+    }
+  }
```

Comment:
It looks to me like this for loop is very similar to the one in `restricted_gibbs_sampling`, am I right? Isn't it possible to avoid redoing this computation and saving all these stuff when you call `restricted_gibbs_sampling()`?

Reply:
Yes, this computation is very similar to `restricted_gibbs_sampling`, with the unique difference that, for each point, the new cluster is not sampled, but it is the cluster of the original configuration.
This step must be done after the restricted Gibbs sampling cycles, so it can't be done during them.

From the point of view of the code, we added a parameter to `restricted_gibbs_sampling` in order to do the computations in the lines above: in this way in `merge` we just need to do a call to `restricted_gibbs_sampling`.
After this change, observing the functions `split` and `merge` we noted that they are nearly identical. 
Is it ok if we remove these two functions and move their code inside `split_or_merge`, using a parameter to identify if we are in the split or in the merge case? In this way, we could avoid duplicated code.

Reply:
That would be great